src/index.ts

```diff
@@ -207,25 +211,79 @@ export async function init(config: IClientConfig): Promise<IEppoClient> {
       throwOnFailedInitialization: true, // always use true here as underlying instance fetch is surrounded by try/catch
       skipInitialPoll: config.skipInitialRequest ?? false,
     };
+    EppoJSClient.instance.setConfigurationRequestParameters(requestConfiguration);
 
-    EppoJSClient.instance.setLogger(config.assignmentLogger);
+    // We have two at-bats for initialization: from the configuration store and from fetching
+    // We can resolve the initialization promise as soon as either one succeeds
+    let initFromConfigStoreError = undefined;
+    let initFromFetchError = undefined;
 
-    // default behavior is to use a LocalStorage-based assignment cache.
-    // this can be overridden after initialization.
-    EppoJSClient.instance.useCustomAssignmentCache(new LocalStorageAssignmentCache());
-    EppoJSClient.instance.setConfigurationRequestParameters(requestConfiguration);
-    await EppoJSClient.instance.fetchFlagConfigurations();
+    const attemptInitFromConfigStore = configurationStore
+      .init()
+      .then(() => {
+        if (!configurationStore.getKeys().length) {
+          // Consider empty configuration stores invalid
+          console.warn('Empty cached configuration');
+          initFromConfigStoreError = new Error('Configuration store was empty');
+          return '';
+        }
+        return 'config store';
+      })
+      .catch((e) => {
+        console.warn('Eppo SDK encountered an error initializing from the configuration store', e);
+        initFromConfigStoreError = e;
+      });
+    const attemptInitFromFetch = EppoJSClient.instance
+      .fetchFlagConfigurations()
+      .then(() => {
+        return 'fetch';
+      })
+      .catch((e) => {
+        console.warn('Eppo SDK encountered an error initializing from fetching', e);
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        initFromFetchError = e;
+      });
+
+    console.log('>>>>> ABOUT TO RACE!');
+
+    let initializationSource = await Promise.race([
```

Comment:
[`Promise.race()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race) will continue execution as soon as the first promise resolves (successfully or unsuccessfully)

This let's us (potentially) resolve initialization as soon as the cache or fetch finishes without waiting for the other.

Reply:
Or is this to prevent some misuse of init where a customer is spamming us needlessly?

Reply:
Yeah a bit of history here, we used to just pass in a (synchronous) configuration store to the common client. For the client SDK, this was backed by local storage. There are some configuration options in the common SDK regarding what to do if the initial request fails. This let clients do things like fall back on serving whatever was in local storage. Some customers would `await` and others wouldn't. This also let them serve stale stuff until fresh stuff came in.

However, two things changed:
1. We have customers who want _asynchronous_ persistent stores (thus the introduction of the hybrid store)
2. We have customers who want finer control over when fetches happen, when cached data is used, etc. (thus this PR)

Some customers care about consistency. Others--such as retailers where there is a direct line from seconds of loading time to revenue--care about speed. The recent work had us biasing towards the former while other customers (e.g., clerk) were asking about the latter. One of the consistency concerns is avoiding layout shift, this mean a customer may want to roll with an expired configuration and just use the updated one next time.